daemon/containerd/resolver.go

```diff
@@ -1,35 +1,89 @@
-		if cfgHost == registry.IndexHostname {
-			cfgHost = registry.DefaultRegistryHost
+func (i *ImageService) newResolverFromAuthConfig(authConfig *registrytypes.AuthConfig) (remotes.Resolver, docker.StatusTracker) {
+	tracker := docker.NewInMemoryTracker()
+	hostsFn := i.registryHosts.RegistryHosts()
+
+	hosts := hostsWrapper(hostsFn, authConfig, i.registryService)
+
+	return docker.NewResolver(docker.ResolverOptions{
+		Hosts:   hosts,
+		Tracker: tracker,
+	}), tracker
+}
+
+func hostsWrapper(hostsFn docker.RegistryHosts, authConfig *registrytypes.AuthConfig, regService registry.Service) docker.RegistryHosts {
+	return docker.RegistryHosts(func(n string) ([]docker.RegistryHost, error) {
+		hosts, err := hostsFn(n)
+		if err != nil {
+			return nil, err
 		}
-		authorizer := docker.NewDockerAuthorizer(docker.WithAuthCreds(func(host string) (string, string, error) {
-			if cfgHost != host {
-				logrus.WithField("host", host).WithField("cfgHost", cfgHost).Warn("Host doesn't match")
-				return "", "", nil
-			}
-			if authConfig.IdentityToken != "" {
-				return "", authConfig.IdentityToken, nil
+
+		for idx, host := range hosts {
+			if host.Authorizer == nil {
+				var opts []docker.AuthorizerOpt
+				if authConfig != nil {
+					opts = append(opts, authorizationCredsFromAuthConfig(*authConfig))
+				}
+				host.Authorizer = docker.NewDockerAuthorizer(opts...)
+
+				isInsecure := regService.IsInsecureRegistry(host.Host)
+				if host.Client.Transport != nil && isInsecure {
+					host.Client.Transport = httpFallback{super: host.Client.Transport}
+				}
+				hosts[idx] = host
 			}
-			return authConfig.Username, authConfig.Password, nil
-		}))
+		}
+		return hosts, nil
+	})
+}
 
-		opts = append(opts, docker.WithAuthorizer(authorizer))
+func authorizationCredsFromAuthConfig(authConfig registrytypes.AuthConfig) docker.AuthorizerOpt {
+	cfgHost := registry.ConvertToHostname(authConfig.ServerAddress)
+	if cfgHost == registry.IndexHostname {
+		cfgHost = registry.DefaultRegistryHost
 	}
 
-	return docker.NewResolver(docker.ResolverOptions{
-		Hosts: docker.ConfigureDefaultRegistries(opts...),
+	return docker.WithAuthCreds(func(host string) (string, string, error) {
+		if cfgHost != host {
+			logrus.WithField("host", host).WithField("cfgHost", cfgHost).Warn("Host doesn't match")
+			return "", "", nil
+		}
+		if authConfig.IdentityToken != "" {
+			return "", authConfig.IdentityToken, nil
+		}
+		return authConfig.Username, authConfig.Password, nil
 	})
 }
+
+type httpFallback struct {
+	super http.RoundTripper
+}
+
+func (f httpFallback) RoundTrip(r *http.Request) (*http.Response, error) {
+	resp, err := f.super.RoundTrip(r)
+	if err != nil {
+		if strings.Contains(err.Error(), "http: server gave HTTP response to HTTPS client") {
+			plainHttpUrl := *r.URL
+			plainHttpUrl.Scheme = "http"
+
+			plainHttpRequest := *r
+			plainHttpRequest.URL = &plainHttpUrl
+
+			return http.DefaultTransport.RoundTrip(&plainHttpRequest)
```

Comment:
https://github.com/golang/go/wiki/CodeReviewComments#initialisms
https://pkg.go.dev/net/http#Request.Clone

```suggestion
			plain := r.Clone(r.Context())
			plain.URL.Scheme = "http"
			return http.DefaultTransport.RoundTrip(plain)
```